clarify sign-in with SSB challenge codes

staltz opened the pull request:
Requesting feedback from @cryptix and @keks, specifically:

- Is this secure?
- What do you think about "hexadecimals" mentioned? Should it be base64? I don't have strong preferences here
- What's a good byte length for `cc` and `sc`?

staltz committed:
clarify sign-in with SSB challenge codes

keks commented:
The diagram doesn't render for me and I'm not 100% sure how to read the syntax. Can you help me out here?

staltz commented:
@keks Sure

![Screenshot from 2021-02-02 14-59-30](https://user-images.githubusercontent.com/90512/106603672-48105980-6567-11eb-8072-81bddf59b2e8.png)

keks commented:
@staltz 

Okay, I think in general this is a reasonable approach. I'd change up a few things, though:
- the domain separation identifier should not just say "=challenge" but very specifically refer to the context this is used in. Something like "=rooms-dashboard-auth" or so.
- Similarly, I think the muxrpc call could also be more specific than "signIn" and make clear that this is a rooms web feature. maybe "rooms.dashboardAuth"?
- Not only one of the challenges should be included, but both, and both IDs, encoded in a way that it is clear what is what

Regarding the encoding:
I'd suggest using a binary encoding instead of hex or base64, but mostly in order to avoid pointless translation between representations. As long as it's consistent, it doesn't matter. Maybe take a look at the SLP encoding from envelope/private groups, I think that does a very similar thing and we could use it here, too: https://github.com/ssbc/envelope-spec/blob/master/encoding/slp.md
To work with this, just encode an array where the first element is the purpose and the rest are the nonces and public keys, always in the same order (e.g. `[ "rooms-dashboard-auth", id_c, id_s, nonce_c, nonce_s]`). This way, the position in array determines meaning. Note that in order to reliably differentiate this from ssb messages we'd have to make sure this doesn't start with a `'{'` byte, the reason we have prepended "=" before, so maybe prepend the "=" before the SLP encoded values. We could also use a different character and reserve "=" for things that are text-representable? Maybe a zero byte, to indicate that it is binary encoding v0?

Regarding the lengths, each of them should be 128 bit. I don't see a possible birthday attack here and we also don't care about quantum computers (in each case we'd have to double the number).

I kind of rushed writing this down because I am already late for something else so please ask of anything is unclear!

staltz commented:
> the domain separation identifier should not just say "=challenge" but very specifically refer to the context this is used in. Something like "=rooms-dashboard-auth" or so.
 
Will do, but what if we'd like to use this same sign-in system in the future for pubs as well? Then the word "rooms" there in the middle isn't good. Note, the different between pub and room is merely which features do they have, and it's entirely possible that some server would have features of both pubs and rooms, thus a hybrid server. How about `=web-dashboard-auth`?

> Similarly, I think the muxrpc call could also be more specific than "signIn" and make clear that this is a rooms web feature. maybe "rooms.dashboardAuth"?

:+1: I'll still think about where to namespace this RPC, but the name will probably be `webDashboardAuth` to reflect the challenge label.

> Not only one of the challenges should be included, but both, and both IDs, encoded in a way that it is clear what is what

You mean that the **signed string** should include all of these? Sounds decent, like we did with alias registration signatures. 

> I'd suggest using a binary encoding instead of hex or base64, but mostly in order to avoid pointless translation between representations.

The challenge nonce would be sent via HTTP params, so I'm not sure there is a benefit of using SLP in an HTTP param. We should use an encoding that is commonly employed for tokens in HTTP params, right?

Also, for alias registration signatures, we don't use SLP, so I'd prefer we use a consistent format for all the signatures in this rooms2 doc.

keks commented:
> Will do, but what if we'd like to use this same sign-in system in the future for pubs as well?

In that case we can see if we really need the same design and either make a new standard that generalizes the two (just like you suggested) or make a separate one that has "pub" in the text. I'd compare this to the browser-specific CSS classes, where the vendor is still in the name (`-moz-...`). It's kind of a standard, but also kind of experimental. Unifying later is easier than introducing differentiation. I'd apply a similar reasoning to the RPC method name.

> You mean that the signed string should include all of these? Sounds decent, like we did with alias registration signatures.

yes, exactly. 

> The challenge nonce would be sent via HTTP params, so I'm not sure there is a benefit of using SLP in an HTTP param. We should use an encoding that is commonly employed for tokens in HTTP params, right?

You can just use URL-safe base64 here.

> Also, for alias registration signatures, we don't use SLP, so I'd prefer we use a consistent format for all the signatures in this rooms2 doc.

Oh, okay. Well since in this case the sizes are of the fields are constant we can just concatenate them. Or do the same as the format in the alias registration. BTW I also feel like the alias registration signature domain separation string ("=alias-registration") could/should mention that this a room2 thing. Sorry for not bringing that up earlier...

---

Also I'd like to contradict past-me and suggest to increase the nonce length to 256 bits. I don't see an attack right now but it looks like a cheap change and I may be overlooking something.

staltz commented:
> BTW I also feel like the alias registration signature domain separation string ("=alias-registration") could/should mention that this a room2 thing. Sorry for not bringing that up earlier...

How about `=room-alias-registration`? There isn't a "rooms 1.0 alias registration" to conflict with.

> Also I'd like to contradict past-me and suggest to increase the nonce length to 256 bits. I don't see an attack right now but it looks like a cheap change and I may be overlooking something.

:+1:

keks commented:
Sounds good!

staltz committed:
Improve sign-in spec with kek's feedback

staltz committed:
sign-out spec

staltz commented:
Updated this PR. Here are the updated UML diagrams (note I added also a signOut spec)

![Screenshot from 2021-02-03 13-53-49](https://user-images.githubusercontent.com/90512/106743638-5756dc00-6627-11eb-994c-f1aacf1f134b.png)

![Screenshot from 2021-02-03 13-53-58](https://user-images.githubusercontent.com/90512/106743657-5aea6300-6627-11eb-9323-7ac8f15edce4.png)

staltz committed:
update the appendix with muxrpc APIs

staltz commented:
@arj03 I just realized that we may want to allow ssb-browser to perform sign-in too. Could you take a look at the UML diagrams, if you see any problem? With ssb-browser, the "SSB peer" and "Browser client" are the same thing.

staltz commented:
I'll just merge this because it's received enough feedback to be considered good enough, but I'm certainly open to more feedback in case y'all have more. Feel free to comment here or in other issues
